### PR TITLE
Shadowmaps

### PR DESCRIPTION
**Pull request by nurgan**
Added final working version of shadow Mapping.
- uses PCF soft shadows (for each fragment only the corresponding shadow map texel is evaluated (resulting in the hard shadow borders, as in the workshop in class), but also the surrounding texels are evaluated, creating a soft shadow outline)
- added a different far plane value for the actual view frustum and for view frustum culling + shadow mapping
- only remaining issue: view frustum still that big, that the shadow map resolution seems to be very low for the size

### CODE REVIEW

File: resources/phong_frag.glsl [hunk 1]
```diff
@@ -34,8 +35,41 @@ uniform float MatShiny;
 uniform mat4 M;
 uniform mat4 V;
 
+uniform vec2 shadowMapSize;
+
+uniform float PCFkernelRadius = 4.0;
+
+//Poisson Disc for PCF filtering
+uniform vec2 poission[25] = { vec2(-0.978698, -0.0884121),
```
Comment (nurgan) [inline resources/phong_frag.glsl:43]: Poission Disc Values to evaluate only some texels in the PCF filter kernel, not all. Precomputed for max render speed.

Comment (rgarmsen2295) [reply resources/phong_frag.glsl:43]: Def curious how this all works as a sweeping comment for the entire PR.  Hopefully, we have time on thursday for a brief overview (this looks good tho).

File: resources/phong_frag.glsl [hunk 2]
```diff
@@ -80,6 +114,51 @@ vec3 dirLightColor(vec3 fragNormal, vec3 view) {
...
+                shadedFrags++;
+            }
+        }
+    }
+
+    // calculate shadow factor
+	float PCFfactor = shadedFrags / samples;
+	return PCFfactor * 0.5 + (1.0 -PCFfactor) * 1.0;
+}
+
```
Comment (rgarmsen2295) [inline resources/phong_frag.glsl:161]: The bits I understand of this look good

File: src/AimRenderComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+void AimRenderComponent::draw(std::shared_ptr<MatrixStack> P, std::shared_ptr<MatrixStack> M, std::shared_ptr<MatrixStack> V) {
+    ShaderManager& shaderManager = ShaderManager::instance();
+    shaderManager.renderObject(holder_, shaderName_, shape_, material_, P, V, M);
+}
+
+void AimRenderComponent::renderShadow(std::shared_ptr <MatrixStack> M) {
+    // don't do anything, the aim shouldn't create a shadow
+}
```
Comment (rgarmsen2295) [inline src/AimRenderComponent.cpp:20]: Me gusta

File: src/main.cpp [hunk 4]
```diff
@@ -77,10 +78,14 @@ int main(int argc, char **argv) {
...
 
+    // Set the Shadow Map Object
+    ShadowMap* shadowMap = new ShadowMap();
+    gameManager.setShadowMap(shadowMap);
+
     // Add all static objects before this!!!
     world.init();
 
-    gameManager.setTime(15.0);
+    gameManager.setTime(1500.0);
```
Comment (rgarmsen2295) [inline src/main.cpp]: I'm fine with increasing the time but just want to check this wasn't left over from debugging unintentionally

**Review by rgarmsen2295 — APPROVED** [review]
Overall, looks good, but, hopefully, we can do a quick run through tomorrow to go through it together in person, I wanna learn about all the cool stuff that's going on!

**Pull request merged**